plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +603,108 @@ public boolean rename(Path src, Path dst)
+            deleteObject(key + DIRECTORY_SUFFIX);
         }
-        catch (FileNotFoundException e) {
+        else {
+            if (path.getName().isEmpty()) {
+                return false; // bucket root does not exist
+            }
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromRecursiveListing));
+            if (objectKeysIterator.hasNext()) {
+                String directoryKey = objectKeysIterator.next();
+                if (!Objects.equals(directoryKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
+            }
+            deleteObject(key + DIRECTORY_SUFFIX);
+        }
+        return true;
+    }
+
+    private boolean deletePrefix(Path prefix)
+    {
+        String bucketName = getBucketName(uri);
+        Iterator<ListObjectsV2Result> listings = listObjects(prefix, OptionalInt.empty(), true);
+        Iterator<String> objectKeys = Iterators.concat(Iterators.transform(listings, TrinoS3FileSystem::keysFromRecursiveListing));
+        Iterator<List<String>> objectKeysBatches = Iterators.partition(objectKeys, BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE);
+        if (!objectKeysBatches.hasNext()) {
             return false;
         }
 
-        if (!recursive) {
-            throw new IOException("Directory " + path + " is not empty");
+        Set<String> objectKeysFailedToDelete = new HashSet<>();
+        while (objectKeysBatches.hasNext()) {
+            String[] objectKeysBatch = objectKeysBatches.next().toArray(String[]::new);
+            try {
+                DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName).withKeys(objectKeysBatch);
+                if (requesterPaysEnabled) {
+                    // TODO use deleteObjectRequest.setRequesterPays() when https://github.com/aws/aws-sdk-java/issues/1219 is fixed
+                    // currently the method exists, but is ineffective (doesn't set the required HTTP header)
+                    deleteObjectsRequest.putCustomRequestHeader(REQUESTER_PAYS_HEADER, REQUESTER_PAYS);
+                }
+                s3.deleteObjects(deleteObjectsRequest);
+            }
+            catch (MultiObjectDeleteException e) {
+                e.getErrors().stream().map(MultiObjectDeleteException.DeleteError::getKey).forEach(objectKeysFailedToDelete::add);
+            }
         }
 
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        // Clean up eventual objects which the deletion has failed while doing the batched deletion
+        while (!objectKeysFailedToDelete.isEmpty()) {
```

Comment:
Why would we retry (twice) here?

Reply:
I think I've overengineered  a bit here. Dropping the retry mechanism altogether.